File: classes.js
class Story {
    constructor(title, description, moral_lesson, ageGroup) {
    this.title = title;
    this.text = description;
    this.moral = moral_lesson;
    this.ageGroup = ageGroup;
    }
    }
    class StoryTeller {
    constructor(name, title, language) {
    this.name = name;
    this.title = title;
    this.language = language;
    }
    tellStories () {
    return `${this.name} is telling ${this.title} in ${this.language}`;
    }
    
    
    };
    let story= new StoryTeller("Grace Ogot","Tekayo","Luo")
    console.log(story.tellStories());
    class Translator extends StoryTeller {
    constructor(name, title, language,targetLanguage) {
    super()
    this.name= name;
    this.title=title;
    this.language=language;
    this.targetLanguage= targetLanguage;
    
    
    }
    
    
    tellStory(){
    
    
    return ` ${this.name} translates ${this.title } from ${this.language} to ${this.targetLanguage} `
    }
    
    
    }
    const translate= new Translator ("Paula Karimi","Tekayo","Luo","Kiswahili")
    console.log(translate.tellStory()) ;
    
    // African Cuisine:** You're creating a recipe app specifically for African cuisine.
    // The app needs to handle recipes from different African countries, each with its
    // unique ingredients, preparation time, cooking method, and nutritional
    // information. Consider creating a `Recipe` class, and think about how you might
    // create subclasses for different types of recipes (e.g., `MoroccanRecipe`,
    // `EthiopianRecipe`, `NigerianRecipe`), each with their own unique properties and
    // Methods.
    
    
    
    class Recipe {
      constructor(name, country, unique_ingredients, preparation_time, cooking_method, nutritional_value) {
        this.name = name;
        this.country = country;
        this.unique_ingredients = unique_ingredients;
        this.preparation_time = preparation_time;
        this.cooking_method = cooking_method;
        this.nutritional_value = nutritional_value;
      }
    }
     class MoroccanRecipe extends Recipe {
      constructor(name, country, unique_ingredients, preparation_time, cooking_method, nutritional_value) {
        super(name, country, unique_ingredients, preparation_time, cooking_method, nutritional_value);
      }
       cook() {
        return `For ${this.name} in ${this.country}, cook the meal with ${this.unique_ingredients} for ${this.preparation_time} by ${this.cooking_method} to gain more ${this.nutritional_value}`;
      }
    }
     class EthiopianRecipe extends Recipe {
      constructor(name, country, unique_ingredients, preparation_time, cooking_method, nutritional_value) {
        super(name, country, unique_ingredients, preparation_time, cooking_method, nutritional_value);
      }
       cook() {
        return `For ${this.name} in ${this.country}, cook the meal with ${this.unique_ingredients} for ${this.preparation_time} by ${this.cooking_method} to gain more ${this.nutritional_value}`;
      }
    }
     class NigerianRecipe extends Recipe {
      constructor(name, country, unique_ingredients, preparation_time, cooking_method, nutritional_value) {
        super(name, country, unique_ingredients, preparation_time, cooking_method, nutritional_value);
      }
       cook() {
        return `For ${this.name} in ${this.country}, cook the meal with ${this.unique_ingredients} for ${this.preparation_time} by ${this.cooking_method} to gain more ${this.nutritional_value}`;
      }
    }
    const moroccan = new MoroccanRecipe("Wheat", "Morocco", "vinegar", "30 minutes", "grilling", "cumin");
    const ethiopian = new EthiopianRecipe("chicken_breasts", "Ethiopia", "white wine", "45 minutes", "stewing", "doro wat");
    const nigerian = new NigerianRecipe("Jollof", "Nigeria", "pepper", "3 hours", "stirring", "proteins");
     console.log(nigerian.cook());
    console.log(moroccan.cook());
    console.log(ethiopian.cook());
    
    
    
    
    
    
    //3. **Wildlife Preservation:** You're a wildlife conservationist working on a
    // program to track different species in a national park. Each species has its own
    // characteristics and behaviors, such as its diet, typical lifespan, migration
    // patterns, etc. Some species might be predators, others prey. You'll need to
    // create classes to model `Species`, `Predator`, `Prey`, etc., and think about how
    // These classes might relate to each other through inheritance.
    
    
    class Species {
     constructor(diet, typicalLifespan, migrationPatterns) {
      this.diet = diet;
      this.typicalLifespan = typicalLifespan;
      this.migrationPatterns = migrationPatterns;
    }
     typeOfAnimal() {
      if (this.diet === "herbivorous") {
        console.log("This animal is not a danger to other animals.");
      } else if (this.diet === "omnivorous") {
        console.log("This animal feeds on plants but also feeds on some animals.");
      } else {
        console.log("This animal is very dangerous to other animals and does not eat plants.");
      }
    }
    }
    class Predator extends Species {
    constructor(diet, typicalLifespan, migrationPatterns, typeOfTeeth, claws, venom, name) {
      super(diet, typicalLifespan, migrationPatterns);
      this.typeOfTeeth = typeOfTeeth;
      this.claws = claws;
      this.venom = venom;
      this.name = name;
    }
     fastKillers() {
      if (this.venom) {
        console.log(`The bite of a ${this.name} can be lethal.`);
      } else {
        console.log(`The bite of a ${this.name} is not lethal.`);
      }
    }
     methodOfKilling() {
      if (this.typeOfTeeth.includes("carnassial teeth")) {
        console.log(`A ${this.name} kills by slicing up their prey.`);
      } else {
        console.log("This predator cannot shear their prey's meat after attacks.");
      }
    }
    }
    class Prey extends Species {
    constructor(diet, typicalLifespan, migrationPatterns, defenseMechanisms, name) {
      super(diet, typicalLifespan, migrationPatterns);
      this.defenseMechanisms = defenseMechanisms;
      this.name = name;
    }
     addDefenseMechanism(defenseMechanism) {
      this.defenseMechanisms.push(defenseMechanism);
    }
    }
    const pred = new Species("grass", "40years", "easttowest");
    pred.typeOfAnimal();
    const deer = new Prey('herbivore', 10, 'seasonal', ['camouflage'], 'Deer');
    console.log(deer.defenseMechanisms);
    const lion = new Predator('carnivore', 15, 'non-migratory', ['carnassial teeth'], 'sharp claws', false, 'Lion');
    lion.methodOfKilling();
    lion.fastKillers();
    deer.addDefenseMechanism('speed');
    console.log(deer.defenseMechanisms);


// 5. Create a class called Product with attributes for name, price, and quantity.
// Implement a method to calculate the total value of the product (price * quantity).
// Create multiple objects of the Product class and calculate their total values.


    class Product {
      constructor(name, price, quantity) {
        this.name = name;
        this.price = price;
        this.quantity = quantity;
      }
     
      calculateTotalValue() {
        if (this.quantity > 0) {
          return this.price * this.quantity;
        } else {
          return 0;
        }
      }
    }
     const products = [
      new Product("pen", 10.5, 100),
      new Product("Book", 0.5, 150),
      new Product("sharpener", 0.2, 120)
    ];
     let totalValue = 0;
     for (let i = 0; i < products.length; i++) {
      if (products[i].quantity > 0) {
        totalValue += products[i].calculateTotalValue();
      }
    }
     console.log("Total value of all products:", totalValue);    

File: classes.kt
// 1.**Ancestral Stories:** In many African cultures, the art of storytelling is passed
// down from generation to generation. Imagine you're creating an application that records these oral stories and translates them into different languages. The
// stories vary by length, moral lessons, and the age group they are intended for.
// Think about how you would model `Story`, `StoryTeller`, and `Translator`
// objects, and how inheritance might come into play if there are different types of
// stories or storytellers.



fun main() {




    val recipes = Recipe(   arrayOf("Rice", "Spices"), "one hour",
        "fry the meat till golden under low heat", "proteins and energy")
    val ethiopianRecipe = EthiopianRecipe("grounded wheat", arrayOf("Wheat", "brown wheat"),
        " am maximum of three days",
        "ferment the flour the fry it under low heat for two minute",
        "energy"
    )
    ethiopianRecipe.ethiopiansSpecial()



   val recipes = Recipe(   arrayOf("Rice", "Spices"), "one hour",
       "fry the meat till golden under low heat", "proteins and energy")
   val ethiopianRecipe = EthiopianRecipe("grounded wheat", arrayOf("Wheat", "brown wheat"),
       " am maximum of three days",
       "ferment the flour the fry it under low heat for two minute",
       "energy"
   )
   ethiopianRecipe.ethiopiansSpecial()




 
 
 }
 
 
 
 
 open class Recipe(var ingredients: Array<String>, var preparationTime: String, val cookingMethod: String, var nutritionalValue: String) {
 
 
    fun getIngredients() {

        println("The ingredients is $ingredients")

 
 
    }
 fun checkallergerns(){
    if (allergens == null)
    allergens=arrayOf<String>(){
    else{
        print(allergens[0])
        }
 }


    fun getPreparationTime() {
        var preptime=10
      if(preptime<=10){
        println("It is an easy meal to prepare")
      }
      else{
        println("The meal will take a longe period to prepare")
      }

      

 
 
    }
 
 
 
    fun servingportions(var guests:Int){
        var foodcapacity =100
        var foodserved = 0
        if(guests <= foodcapacity){
            foodserved = guests
            }else{
                foodserved = foodcapacity



        }
       
    }
       fun portions(preferredServing: Int): String {
       return when {
           preferredServing > serving -> {
               val portionAmount = (preferredServing / serving)
               "The portion of ingredients can be increased $portionAmount times"
           }
                   preferredServing < serving -> {
               val portionAmount = (serving / preferredServing)
               "The portion of ingredients can be decreased $portionAmount times"
           }
           else ->{
               "The portion of food serves $serving"
           }
       }
   }
 
 }
 
 
 class MorrocanRecipe(var specialIngredient: String, ingredients: Array<String>, preparationTime: String, cookingMethod: String,
    
 nutritionalValue: String) : Recipe(ingredients, preparationTime, cookingMethod, nutritionalValue) {
    fun specialRecipe() {
        println("The special recipe of Morocco is $specialIngredient")
 
 
    }
 
 
 }
 
 
 
 
 class EthiopianRecipe(var specialIngredient: String, ingredients: Array<String>, preparationTime: String, cookingMethod: String,
    nutritionalValue: String) : Recipe(ingredients, preparationTime, cookingMethod, nutritionalValue) {
 
 
    fun ethiopiansSpecial() {
        println("The special recipe of Ethiopia is $specialIngredient")
 
 
 
 
    }
 
 
 }
 
 
 class NigerianRecipe(var specialIngredient: String, ingredients: Array<String>, preparationTime: String, cookingMethod: String,
    nutritionalValue: String) : Recipe(ingredients, preparationTime, cookingMethod, nutritionalValue) {
 
 
    fun nigerianSpecial() {
        println("The special recipe of Nigeria is $specialIngredient ")
       
    }
 }





 //
// 2. **African Cuisine:** You're creating a recipe app specifically for African cuisine.
// The app needs to handle recipes from different African countries, each with its
// unique ingredients, preparation time, cooking method, and nutritional
// information. Consider creating a `Recipe` class, and think about how you might
// create subclasses for different types of recipes (e.g., `MoroccanRecipe`,
// `EthiopianRecipe`, `NigerianRecipe`), each with their own unique properties and
// Methods.


class MorrocanRecipe(var specialIngredient: String, ingredients: Array<String>, preparationTime: String, cookingMethod: String,
   
nutritionalValue: String) : Recipe(ingredients, preparationTime, cookingMethod, nutritionalValue) {
   fun specialRecipe() {
       println("The special recipe of Morocco is $specialIngredient")


   }


}




class EthiopianRecipe(var specialIngredient: String, ingredients: Array<String>, preparationTime: String, cookingMethod: String,
   nutritionalValue: String) : Recipe(ingredients, preparationTime, cookingMethod, nutritionalValue) {


   fun ethiopiansSpecial() {
       println("The special recipe of Ethiopia is $specialIngredient")




   }


}


class NigerianRecipe(var specialIngredient: String, ingredients: Array<String>, preparationTime: String, cookingMethod: String,
   nutritionalValue: String) : Recipe(ingredients, preparationTime, cookingMethod, nutritionalValue) {


   fun nigerianSpecial() {
       println("The special recipe of Nigeria is $specialIngredient ")
      
   }
}


// 3. **Wildlife Preservation:** You're a wildlife conservationist working on a
// program to track different species in a national park. Each species has its own
// characteristics and behaviors, such as its diet, typical lifespan, migration
// patterns, etc. Some species might be predators, others prey. You'll need to
// create classes to model `Species`, `Predator`, `Prey`, etc., and think about how
// These classes might relate to each other through inheritance.

fun main() {
    val animal1 = Predator("Leopard", "meat", "30yrs", "East to West", "spot and stalk")
    println(animal1.speciesInfo())
    val animal2 = Prey("Antelope", "grass", "20yrs", "seasonal", "lying flat on ground")
    println(animal2.speciesInfo())
   val animal3=Animal("Lion","Meat","50yrs","West
    to East","Roar")
 
 
 
 
 }
 
 
 
 
 open class Species(val name: String, val diet: String, val lifespan: String, val migrationPatterns: String) {
   open fun speciesInfo(): String {
        return "$name feeds on $diet, it has a lifespan of $lifespan and migrates $migrationPatterns"
    }
 }
 class Predator(name: String, diet: String, lifespan: String, migrationPatterns: String, val huntingMethod: String) : Species(name, diet, lifespan, migrationPatterns) {
    override fun speciesInfo(): String {
        return super.speciesInfo()
    }
    
 }
 
 
 class Prey(name: String, diet: String, lifespan: String, migrationPatterns: String, val defenseMechanism: String) : Species(name, diet, lifespan, migrationPatterns) {
    override fun speciesInfo(): String {
        return super.speciesInfo()
    }



4.**African Music Festival:** You're in charge of organizing a Pan-African music
festival. Many artists from different countries, each with their own musical style
and instruments are scheduled to perform. You need to write a program to
manage the festival lineup, schedule, and stage arrangements. Think about how
you might model the `Artist`, `Performance`, and `Stage` classes, and consider
how you might use inheritance if there are different types of performances or stages.








open class Artist(val name: String, val country: String, val musicType: String, val instruments: List<String>) {
    fun artistDetail() {
        println("name: $name, country: $country, musicType: $musicType, instruments: ${instruments.joinToString(", ")}")
    }
}
class Performance(name: String, country: String, musicType: String, instruments: List<String>, val stageTime: String) :
    Artist(name, country, musicType, instruments) {
    fun scheduling() {
        println("name: $name, musicType: $musicType, time: $stageTime")
    }
    fun playInstrument(instrument: String) {
        if (instruments.contains(instrument)) {
            println("$name is playing $instrument")
        } else {
            println("$name is not playing $instrument")
        }
    }
}

fun main() {
    val artist = Artist("jayz", "Canada", "hiphop", listOf("guitar", "piano"))
    artist.artistDetail()
    val perform = Performance("sautisol", "Kenya", "bongo", listOf("guitar", "piano"), "2 hours")
    perform.scheduling()
    perform.playInstrument("guitar")
    val stage = Stage("Bridget blue", "Kenya", "gospel", listOf("piano", "harp", "violin"), 100, "Nakuru")
    stage.performance()
    val artist1 = Artist("Nigeria", "Fela Kuti", listOf("guitar", "drums"))
    val artist2 = Artist("Ghana", "Kofi Annan", listOf("piano", "violin"))
    val artist3 = Artist("Kenya", "Wangari Maathai", listOf("flute", "harp"))
    val performance1 = Performance(artist1.name, artist1.country, artist1.musicType, artist1.instruments, "Friday, 10am")
    val performance2 = Performance(artist2.name, artist2.country, artist2.musicType, artist2.instruments, "Saturday, 2pm")
    val performance3 = Performance(artist3.name, artist3.country, artist3.musicType, artist3.instruments, "Sunday, 4pm")
    val stage1 = Stage("Open-air stage", "The Great Lawn", 10000)






val stage2 = Stage("Indoor stage", "The Concert Hall", 5000)
    println("The festival lineup is:")
    listOf(performance1, performance2, performance3).forEach { performance ->
        println(performance.festivalLineup?.let {
            "name: ${it.name}, country: ${it.country}, musicType: ${it.musicType}, instruments: ${it.instruments.joinToString(", ")}"
        })
    }
}
